safehttp/plugins/csp/csp.go

```diff
@@ -0,0 +1,170 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package csp
+
+import (
+	"context"
+	"crypto/rand"
+	"encoding/base64"
+	"strings"
+
+	"github.com/google/go-safeweb/safehttp"
+)
+
+var randReader = rand.Reader
+
+// nonceSize is the size of the nonces in bytes.
+const nonceSize = 20
+
+func generateNonce() string {
+	b := make([]byte, nonceSize)
+	_, err := randReader.Read(b)
+	if err != nil {
+		// TODO: handle this better, what should happen here?
+		panic(err)
+	}
+	return base64.StdEncoding.EncodeToString(b)
+}
+
+// Policy defines a CSP policy.
+type Policy struct {
+	reportOnly bool
+
+	// serialize serializes this policy for use in a Content-Security-Policy header
+	// or in a Content-Security-Policy-Report-Only header. If the given context
+	// contains a nonce, it is used, otherwise a new one is generated and placed
+	// in the context.
+	serialize func(context.Context) (string, context.Context)
+}
+
+type ctxKey struct{}
+
+// Nonce retrieves the nonce from the given context. If there is no nonce stored
+// in the context, a new nonce is generated and placed in the context.
+func Nonce(ctx context.Context) (string, context.Context) {
+	v := ctx.Value(ctxKey{})
+	if v == nil {
+		v = generateNonce()
+		ctx = context.WithValue(ctx, ctxKey{}, v)
+	}
+	return v.(string), ctx
+}
+
+// StrictCSPBuilder can be used to build a strict, nonce-based CSP.
+// See https://csp.withgoogle.com/docs/strict-csp.html for more info.
+//
+// If BaseURI is an empty string the base-uri directive will be set to 'none'.
+type StrictCSPBuilder struct {
+	ReportOnly    bool
+	StrictDynamic bool
+	UnsafeEval    bool
+	BaseURI       string
+	ReportURI     string
+}
+
+// Build creates a Policy based on the specified options.
+func (s StrictCSPBuilder) Build() Policy {
+	return Policy{
+		reportOnly: s.ReportOnly,
+		serialize: func(ctx context.Context) (string, context.Context) {
+			var b strings.Builder
+
+			b.WriteString("object-src 'none'; script-src 'unsafe-inline' https: http: 'nonce-")
+			n, ctx := Nonce(ctx)
```

Comment:
I don't think `serialize` should be in charge of generating a nonce and updating the context. It should take the nonce as a parameter and `Before` should do the generation and context alteration instead.

Reply:
Also, if you do that, then it's going to be clear that the nonce is not in the context, so you don't even need to check that (remember, it's a linear search).